packages/nextjs/app/builders/page.tsx

```diff
@@ -0,0 +1,122 @@
+
+const BATCH_REGISTRY = externalContracts[arbitrum.id]?.BatchRegistry?.address as Address;
+
+const BatchRegistryAbi = [
+  {
+    type: "event",
+    name: "CheckedIn",
+    inputs: [
+      { name: "first", type: "bool", indexed: false },
+      { name: "builder", type: "address", indexed: false },
+      { name: "checkInContract", type: "address", indexed: false },
+    ],
+    anonymous: false,
+  },
+] as const;
+
+const client = createPublicClient({
+  chain: arbitrum,
+  transport: http(process.env.NEXT_ARB_RPC_URL || "https://arb1.arbitrum.io/rpc"),
+});
+
+function findBuilderPagePath(addr: string): string | null {
+  const variations = [
+    addr.toLowerCase(),
+    addr.toUpperCase(),
+    getAddress(addr), // checksummed
+    addr, // original
+  ];
+
+  for (const variation of variations) {
+    const p = path.join(process.cwd(), "app", "builders", variation, "page.tsx");
+    if (fs.existsSync(p)) {
+      return variation;
+    }
+  }
+
+  return null;
+}
+
+async function getBuildersFromEvents(): Promise<string[]> {
+  const logs = await client.getLogs({
+    address: BATCH_REGISTRY,
+    fromBlock: 0n,
+    toBlock: "latest",
+  });
+
+  const firstTimeBuilders = new Set<string>();
+
+  for (const log of logs) {
+    try {
+      if (!log.topics || log.topics.length === 0) continue;
+
+      const decoded = decodeEventLog({
+        abi: BatchRegistryAbi,
+        data: log.data,
+        topics: log.topics as [signature: `0x${string}`, ...args: `0x${string}`[]],
+      });
+
+      if (decoded.eventName === "CheckedIn") {
+        const { first, builder } = decoded.args as {
+          first: boolean;
+          builder: Address;
+          checkInContract: Address;
+        };
+        if (first) {
+          firstTimeBuilders.add(builder.toLowerCase());
+        }
+      }
+    } catch {}
+  }
+  return Array.from(firstTimeBuilders).sort();
+}
+
+export default async function BuildersPage() {
+  const builders = await getBuildersFromEvents();
+
+  return (
+    <main className="mx-auto max-w-3xl p-6">
+      <h1 className="text-2xl font-bold mb-1">Batch #20 Builders</h1>
+      <p className="text-sm text-zinc-100 mb-6">
+        Fetched from <code className="font-mono">{BATCH_REGISTRY}</code> (Arbitrum) via <code>CheckedIn</code> events.
+      </p>
+
+      {builders.length === 0 ? (
+        <p>No builders found yet.</p>
+      ) : (
+        <ul className="space-y-2">
+          {builders.map(addr => {
+            const resolvedPagePath = findBuilderPagePath(addr);
+            const short = `${addr.slice(0, 6)}…${addr.slice(-4)}`;
+
+            return (
+              <li key={addr} className="border p-3 rounded-xl flex items-center justify-between">
+                {resolvedPagePath ? (
+                  <a className="font-mono underline" href={`/builders/${resolvedPagePath}`}>
+                    {short}
+                  </a>
+                ) : (
+                  <span className="font-mono">{short}</span>
```

Comment:
There's a component in this project that renders the address. You should try using it.